### ISSUE TITLE
Conan2: how to set package root folder while creating a package

### ISSUE THREAD

**Issue by FaiqueAli** [conan-io/conan#16738]
### What is your question?

I am trying to create a `conan2` package using `conan create` command. To compile the project, I am calling a bash script that should run from the root folder of the project. I have noticed that my script is not found `/bin/sh: 1: run-linux.sh: not found` during the package creation `conan create . -s build_type=Release  -pr:b=gcc_build -pr:h=gcc_build --build missing`, because the current directory is `User/.conan2/p/b/cucum1a6bb099e2a62/b/build/Release`.
how can I set the Conan hash root folder `User/.conan2/p/b/cucum1a6bb099e2a62/b` from which all the Conan commands should be executed?
below is the `conanfile.py` for better understanding

```
from conan import ConanFile
from conan.tools.cmake import CMakeToolchain, CMake, cmake_layout, CMakeDeps


class cucumberqnx(ConanFile):
    name = "cucumberqnx"
    version = "1.0"

    # Optional metadata
    license = "<Put the package license here>"
    author = "<xxx> <xxx>"
    url = "<Package recipe repository url here, for issues about the package>"
    description = "<xxx>"
    topics = ("<qnx>", "<cucumber>", "<qcc cucumber>")

    # Binary configuration
    settings = "os", "compiler", "build_type", "arch"
    options = {"shared": [True, False], "fPIC": [True, False]}
    default_options = {"shared": False, "fPIC": True}
    #package_type = "header-library"
    # Sources are located in the same place as this recipe, copy them to the recipe
    exports_sources = "CMakeLists.txt", "src/*", "include/*", "cmake", "run-linux.sh"

    def layout(self):
        cmake_layout(self)
    
    def requirements(self):
        self.requires("asio/1.30.2")
        self.requires("gtest/1.14.0")
        self.requires("tclap/1.2.5")

    def generate(self):
        deps = CMakeDeps(self)
        deps.generate()
        tc = CMakeToolchain(self)
        tc.generate()

    def build(self):
         cmake = CMake(self)
         
    def package(self):
        cmake = CMake(self)
        cmake.install()

    def package_info(self):
        self.cpp_info.libs = ["cucumberqnx"]
```

### Have you read the CONTRIBUTING guide?

- [x] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @FaiqueAli 

Thanks for your question.

It is not very clear where your script is being used, I don't see any use of the ``run-linux.sh`` script in the recipe. If you can provide something that we can reproduce (including all the necessary files and details), that would greatly help. Also, it is very likely that some parts are not necessary, for example the ``requires`` might be completely unrelated

> User/.conan2/p/b/cucum1a6bb099e2a62/b

The folder in which Conan builds the packages in the cache is fixed and cannot be changed. This is the only way to keep the cache structure consistent and safe, avoiding overlaps and overwrites of different configurations and builds. 
I don't fully understand why would you want to change this folder, maybe you can clarify this point?

**Comment by FaiqueAli** [comment]
Hi thanks for the reply,

For normal compilation (without Conan) the script is used to build the project from the root of the project. this script contains all the settings (activate/deactivate options defined in cmake) and build commands to compile the project. This script is provided by the cucumber-cpp repository.The project is being compiled, when I execute only this script on my local machine. 

Now I want to create a conan package based on this cmake project, so that I can use its dependencies in another project which is a Makefile based project used for cross-compilation. 

The problem is, as soon I execute  `conan create . -s build_type=Release  -pr:b=gcc_build -pr:h=gcc_build --build missing`  and when it comes to the generate method the execution is transfered to the `/.conan2/p/cucume206078553ec8/e/build/Release/generate`  folder rather than `/.conan2/p/cucume206078553ec8/e`  folder.

if you would like to review you could have a look at the [cucumber-cpp](https://github.com/cucumber/cucumber-cpp) git repository

**Comment by memsharded** [comment]
I am checking the contents of https://github.com/cucumber/cucumber-cpp/blob/main/run-linux.sh, and so far I'd say that I'd try to avoid this script:

- Most of it is intended to run tests, which wouldn't be that critical for most package creation. ConanCenter recipes and packages skip running tests, as it is expensive, and ConanCenter is for package creation, not for library testing and validation
- The ``cmake`` commands are pretty straightforward, the tools like ``CMakeToolchain`` can be easily leveraged to define the correct cmake variables for the build
- But the most important thing, doing a ``sudo cmake --install build`` (the last line of the script) is not good practice for package management. Using ``sudo`` shouldn't be necessary, and it would be extremely discouraged for most package building. Also, it would make very difficult to separate the ``build()`` part from the ``package()`` part, as it is in the same script.


As I see in the cucumber repo, the ``CMakeLists.txt`` is not calling the script, but it is intended the other way round. But your above ``conanfile.py`` doesn't call the ``run-linux.sh`` script, it will execute directly CMake commands, and I think that is a better way for this particular use case.

**Comment by FaiqueAli** [comment]
Ok, My intension is just to compile the project in a way that the repository is doing.
if I do without the script I also found couple of errors.
first I do not know how can I pass settings to the cmake file

```
Cmake -G Ninja \
-DCUKE_ENABLE_BOOST_TEST=on \
-DCUKE_ENABLE_GTEST=on \
-DCUKE_ENABLE_QT_6=OFF \
-DCUKE_ENABLE_EXAMPLES=on \
-DCUKE_TESTS_UNIT=on \
-DCUKE_CODE_COVERAGE=on \
```
I tried these inside conanfile.py like 

```
def build(self):
cmake = CMake(self)
cmake.definitions["-DCUKE_ENABLE_BOOST_TEST"] = "ON"
cmake.definitions["-DCUKE_ENABLE_GTEST"] = "ON"
cmake.definitions["-DCUKE_ENABLE_QT_6"] = "OFF"
cmake.definitions["-DCUKE_ENABLE_EXAMPLES"] = "ON"
cmake.definitions["-DCUKE_TESTS_UNIT"] = "ON"
cmake.definitions["-DCUKE_CODE_COVERAGE"] = "ON"
cmake.configure()
cmake.build()
```

but I am getting errors

```
cucumberqnx/1.0: ERROR: 
Package 'cea0c809402019f2aab57662448eac835c742320' build failed
cucumberqnx/1.0: WARN: Build folder /.conan2/p/b/cucumed179cc6d8a33/b/build/Release
ERROR: cucumberqnx/1.0: Error in build() method, line 58
        cmake.definitions["-DCUKE_ENABLE_BOOST_TEST"] = "ON"
        AttributeError: 'CMake' object has no attribute 'definitions'
```
could you please guide me how can I proceed ?
I just wanted to compile not to execute tests..

**Comment by memsharded** [comment]
> Ok, My intension is just to compile the project in a way that the repository is doing.

The problem is the script is doing more than compiling/building. It is also running some tests, but the worst part is that it is installing the final binaries in the system with ``sudo`` permission. This shouldn't be done by a normal library or package build.


> cmake.definitions["-DCUKE_ENABLE_BOOST_TEST"] = "ON"

This is legacy Conan 1.X syntax. Please check ``CMakeToolchain`` and using ``cache_variables``: https://docs.conan.io/2/reference/tools/cmake/cmaketoolchain.html#cache-variables

**Comment by memsharded** [comment]
Please note the above values are not great, most likely you should put to OFF all test things, and also the ``coverage`` and ``examples`` building should be OFF, those are time consuming and shouldn't be necessary for a regular package creation.

**Comment by FaiqueAli** [comment]
So, I used the cache_variables way to define it, and I guess it is working. I was also exporting wrongly the `cmake/module` folder from the `conanfile.py ` which I also changed from `exports_sources = "CMakeLists.txt", "src/*", "include/*", "cmake"` to 
`exports_sources = "CMakeLists.txt", "src/*", "include/*", "cmake/*"` in the conanfile.py.

Now I am getting some other error which is I guess related with git versioning please have a look at the output below of my `conan create `command

```
CMake Error at cmake/modules/GitVersion.cmake:18 (message):
  Error running 'git describe': fatal: not a git repository (or any of the
  parent directories): .git

Call Stack (most recent call first):
  src/CMakeLists.txt:130 (git_get_version)


-- Configuring incomplete, errors occurred!

cucumberqnx/1.0: ERROR: 
Package 'cea0c809402019f2aab57662448eac835c742320' build failed
cucumberqnx/1.0: WARN: Build folder /.conan2/p/b/cucum01abdf142fbca/b/build/Release
ERROR: cucumberqnx/1.0: Error in build() method, line 57
        cmake.configure()
        ConanException: Error 1 while executing
```

although I can see the output of the `git status` command in my project folder.

Note: if I create a package from non git repo folder, it is complaining about the `CUKE_VERSION` 


```
./conan2/p/b/cucuma8939da8d2315/b/src/main.cpp: In function ‘int main(int, char**)’:
/.conan2/p/b/cucuma8939da8d2315/b/src/main.cpp:44:56: error: ‘CUKE_VERSION’ was not declared in this scope; did you mean ‘ASIO_VERSION’?
   44 |     TCLAP::CmdLine cmd("C++ Cucumber wireserver", ' ', CUKE_VERSION);
      |                                                        ^~~~~~~~~~~~
      |                                                        ASIO_VERSION
[ 98%] Linking CXX static library libcucumber-cpp-nomain.a
[ 98%] Built target cucumber-cpp-nomain
gmake[2]: *** [src/CMakeFiles/cucumber-cpp.dir/build.make:300: src/CMakeFiles/cucumber-cpp.dir/main.cpp.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
gmake[1]: *** [CMakeFiles/Makefile2:154: src/CMakeFiles/cucumber-cpp.dir/all] Error 2
gmake: *** [Makefile:136: all] Error 2

cucumberqnx/1.0: ERROR:
Package 'cea0c809402019f2aab57662448eac835c742320' build failed
cucumberqnx/1.0: WARN: Build folder /.conan2/p/b/cucuma8939da8d2315/b/build/Release
ERROR: cucumberqnx/1.0: Error in build() method, line 58
        cmake.build()
        ConanException: Error 2 while executing
```

**Comment by memsharded** [comment]
> CMake Error at cmake/modules/GitVersion.cmake:18 (message):
  Error running 'git describe': fatal: not a git repository (or any of the
  parent directories): .git


This is because the build is assuming it is in a git repo, but it is not, because you are using the ``export_source`` approach, that copies the sources, but not the whole repo.

There would be 2 different approaches to this:

- In general, it is not necessary that CMake runs ``git describe`` to get the current version from git, as it can have other mechanisms to define it. You can do a patch in your ``conanfile.py`` to disable this behavior in the library ``GitVersion.cmake`` file.
- Another approach would be to have an actual repo, that could be done using a ``scm`` strategy for sources, instead of using the ``export_source`` approach. Please check https://docs.conan.io/2/examples/tools/scm/git/capture_scm/git_capture_scm.html to read more about the ``scm`` approach

**Comment by FaiqueAli** [comment]
Hi I would like to close this ticket now, because I have moved to another solution for my target task.
Initially, I wanted to create a `conan package ` from [cucumber-cpp](https://github.com/cucumber/cucumber-cpp) repo and then wanted to use that package in Makefile based project for cross compilation(for QNX). Now I found [gurkenlaeufer](https://github.com/paulelsner/gurkenlaeufer) as a [conan package](https://conan.io/center/recipes/gurkenlaeufer?version=1.1.0) and successfully consuming it for cross compilation (to execute BDD tests in QNX).
So, the current thread is not relevent anymore.
But for the future readers, initially, I like to compile my project using a bash script that sets the `CMake` variable using that script. but I figured out that I can also set the variables inside the conanfile.py. 
Below is the example that worked in my first case..

```
 def build(self):
        cmake = CMake(self)
        cmake.configure(variables={"CUKE_ENABLE_BOOST_TEST":False,"CUKE_ENABLE_QT":False,"CUKE_TESTS_UNIT":False})
        cmake.build()
```

I would like to say thanks to Mr @memsharded  for the prompt replies and spending his valueable time for my task.

**Comment by FaiqueAli** [comment]
Hi I would like to close this ticket now, because I have moved to another solution for my target task.
Initially, I wanted to create a `conan package ` from [cucumber-cpp](https://github.com/cucumber/cucumber-cpp) repo and then wanted to use that package in Makefile based project for cross compilation(for QNX). Now I found [gurkenlaeufer](https://github.com/paulelsner/gurkenlaeufer) as a [conan package](https://conan.io/center/recipes/gurkenlaeufer?version=1.1.0) and successfully consuming it for cross compilation (to execute BDD tests in QNX).
So, the current thread is not relevent anymore.
But for the future readers, initially, I like to compile my project using a bash script that sets the `CMake` variable using that script. but I figured out that I can also set the variables inside the conanfile.py. 
Below is the example that worked in my first case..

```
 def build(self):
        cmake = CMake(self)
        cmake.configure(variables={"CUKE_ENABLE_BOOST_TEST":False,"CUKE_ENABLE_QT":False,"CUKE_TESTS_UNIT":False})
        cmake.build()
```

I would like to say thanks to Mr @memsharded  for the prompt replies and spending his valueable time for my task.
